[Kwan Zhi Xuan Jacob] iP

jacobkwan opened the pull request:
# Duke

> Due Today, Do Tomorrow

Duke frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~FAST~ _SUPER_ FAST to use

All you need to do is:
1. Download Duke from [here](https://github.com/jacobkwan/ip/releases/tag/A-Jar)
2. Double click it
3. Add your Todos/Deadlines/Events
4. Sit back and let your task list grow! 👍 💯 🥇 👎 

Features: 
- [x] Managing Tasks
- [ ] Adding Tags
- [ ] Reminders

Get it **NOW** while it is still **FREE**!

For the programmers out there, you can download the source code and add your own features!
Here's the `main` method:
```java
public static void main(String[] args) {
        new Duke().run();
}
```

jacobkwan committed:
Level 1. Greet, Echo, Exit

jacobkwan committed:
Level 2. Add, List

jacobkwan committed:
add support for marking tasks as done and undone

jacobkwan committed:
add support for specific tasks: todos, events, deadlines

jacobkwan committed:
support script based text ui testing

jacobkwan committed:
abstract application logic, add basic error handling

jacobkwan committed:
add support for deleting tasks

jacobkwan committed:
update test cases to validate error handling and delete functionality

jacobkwan committed:
Add support for persisting tasks via txt file

jacobkwan committed:
treat date inputs as LocalDate objects

jacobkwan committed:
Merge branch 'branch-Level-7'
add support for persisting tasks to memory

jacobkwan committed:
Merge branch 'branch-Level-8'
add support for date types, resolve minor merge conflict with imports

jacobkwan committed:
standardise date representation in txt file

jacobkwan committed:
add duke.txt to gitignore

jacobkwan committed:
abstract out each command, add helper classes

jacobkwan committed:
separate various classes into their own packages

jacobkwan committed:
import gradle config files and wrapper

jacobkwan committed:
Add unit tests for Parser and TaskList

jacobkwan committed:
fix regression that allowed blank descriptions

jacobkwan committed:
add versioning for jar file

jacobkwan committed:
add javadocs for majority of classes

jacobkwan committed:
add support for searching by keyword(s)

jacobkwan committed:
follow coding standards

jacobkwan committed:
Merge branch 'branch-A-JavaDoc'

jacobkwan committed:
Merge branch 'branch-A-CodingStandard'

jacobkwan committed:
Merge branch 'branch-Level-9'

jacobkwan committed:
format javadocs

jacobkwan committed:
add javadocs for FindCommand

jacobkwan committed:
follow naming convention for junit tests

Berted commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,110 @@
+     * Creates a TaskList with no tasks stored.
+     */
+    public TaskList() {
+        this.tasks = new ArrayList<>();
```
```suggestion
        tasks = new ArrayList<>();
```

Should the `this` specifier be removed as [it is not shadowed by a method or constructor parameter](https://se-education.org/guides/conventions/java/index.html#variables)?

Berted commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,114 @@
+                description = restOfCommand.substring(0, index);
+                if (description.isBlank()) {
+                    throw new DukeException("Description cannot be empty.");
+                }
```
I like the use of [guard clauses](https://se-education.org/guides/conventions/java/index.html#variables) which increases the readability of your code. Keep it up!

Berted commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,58 @@
+ * Handles displaying content to the user.
+ */
+public class Ui {
+    private Scanner sc = new Scanner(System.in);
```
Perhaps consider renaming `sc` to [increase clarity](https://se-education.org/guides/conventions/java/index.html#naming)?

Berted commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,114 @@
+            try {
+                int index = restOfCommand.lastIndexOf("/at");
+                if (index == -1) {
+                    throw new DukeException("To add a deadline, please specify '/by dd/mm/yy'.");
```
```suggestion
                    throw new DukeException("To add an event, please specify '/by dd/mm/yy'.");
```
This seems to have been a typo. Perhaps you have meant `"to add an event"`?

jacobkwan replied:
yup thanks for the catch!

Berted commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,58 @@
+    public void printString(String input) {
+        // String indentedInput = input.replaceAll("(?m)^", "\t");
+        // System.out.println("\t----------------------------------");
+        System.out.println("----------------------------------");
```
```suggestion
        this.showLine();
```
Should `this.showLine()` be used instead to improve code reuse?

marclzh commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,114 @@
+     * @throws DukeException If fullCommand is of an invalid format.
+     */
+    public static Command parse(String fullCommand) throws DukeException {
+        fullCommand = fullCommand.strip();
```
I like the use of `.strip` to ensure the correctness of the data input!

marclzh commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,29 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task {
```
I think adding javadocs could help make what the methods and class do clearer!

marclzh commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,29 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+public class Event extends Task {
```
Similarly for this class, I think javadocs could be added.

marclzh commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,39 @@
+package duke.task;
+
+public abstract class Task {
```
Similarly for this class, I think javadocs could be added.

marclzh commented on src/main/java/duke/task/Todo.java:
```diff
@@ -0,0 +1,20 @@
+package duke.task;
+
+public class Todo extends Task {
```
Similarly for this class, I think javadocs could be added.

marclzh reviewed (commented):
Overall looks well done with no real issues, except some classes missing javadocs There are some design implementations which I can learn from as well!

Berted reviewed (commented):
LGTM, especially in terms of following the provided coding standard!

jacobkwan committed:
remove unecessary use of `this` modifier, fix minor typos

jacobkwan committed:
add basic GUI functionality

jacobkwan committed:
Merge branch 'branch-Level-10'

jacobkwan committed:
use fx:root for MainWindow instead

jacobkwan committed:
add checkstyle and fix minor formatting issues

jacobkwan committed:
add checkstyle to build.gradle

jacobkwan committed:
update jar version

jacobkwan committed:
add simple assertions

jacobkwan committed:
Add missing javadocs

There are several methods and classes that are completely missing
javadocs.

Adding these missing javadocs improves the code quality by improving
documentation.

Let's add the missing javadocs into our project.

jacobkwan committed:
Merge pull request #2 from jacobkwan/branch-A-Assertions

Add simple assertions

jacobkwan committed:
Merge branch 'master' into branch-A-CodeQuality

jacobkwan committed:
Merge pull request #3 from jacobkwan/branch-A-CodeQuality

Add missing javadocs

jacobkwan committed:
Add batch operations for Delete commands

Users are unable to delete multiple tasks in one command.

Batch operations are more user friendly as users often want to operate
on several tasks at one time.

Let's modify the Delete command to allow users to input comma
separated task IDs that they want to delete.

jacobkwan committed:
Add batch operations for Mark and Unmark commands

Users are unable to mark multiple tasks as done or undone.

Batch operations improve user experience as users often want to do
things as quickly as possible.

Let's modify Mark and Unmark commands to allow users to input comma
separated indices of the tasks they would like to mark as done or
undone.

jacobkwan committed:
Merge branch 'branch-C-MassOps'

jacobkwan committed:
add Ui.png

jacobkwan committed:
Set title for Duke window

jacobkwan committed:
Add User Guide

jacobkwan committed:
formatting changes

jacobkwan committed:
change formatting for User Guide

jacobkwan committed:
Improve abstraction in Parser class

jacobkwan committed:
Fix behaviour for Bye command
